### ISSUE TITLE
Insertion performance dramatically drops with amount of documents in index

### ISSUE THREAD

**Issue by gnur** [meilisearch/meilisearch#900]
**Describe the bug**
Duration of calls to the addorreplace endpoint using the Go SDK scale linearly with the amount of documents in the index.

**To Reproduce**
1. Insert about 200k documents over the course of 2 days
2. Graph insertion time and number of documents
3. See the graphs almost exactly line up

**Expected behavior**
I would've hoped (and expected) that index size itself had a smaller effect on the insertion time. I'm using a simple addorreplace call to insert about 20 - 30 documents at a time. There is another process that needs about 200ms per insertion and I want to make sure that the process is not stalled so I insert them either every 20 seconds or when 50 documents are reached.

I have about 800k documents to index and between 0.12 and 0.13 not being backwards compatible (with a point release of the SDK containing a breaking change) and indexing that will take more then a week I'm considering alternatives right now.

**Screenshots**
<img width="1012" alt="Screenshot 2020-08-07 at 14 56 15" src="https://user-images.githubusercontent.com/722329/89647635-25ee9500-d8be-11ea-9bc5-a986d1460d63.png">

As you can see in this graph, with about 200k documents in the index, the time taken per document was 500ms. While the time per document with an empty index was only 0.4ms.

**Additional context**
It looks like meilisearch is only using a single core for indexing on my 8 core machine.  

Example document:

```
{
   "Hash":"homertheiliad",
   "Title":"The Iliad",
   "Author":"Homer",
   "Language":"en",
   "Description":"Homer wrote a beautiful book called the Iliad a long long time ago",
   "Added":"2020-06-22T10:12:23.915683+02:00",
   "Path":"/var/folders/hv/87t09nds7c53nz7115t6fkn43d0z81/T/tmp.tDWsBVXZ/H/Homer/Homer-The_Iliad.epub"
}
```
Hash is the primary key for the index and is unique among insertions so no replacements happen.

I don't know if this is an issue with how I use meilisearch or with meilisearch itself but if there is anything I can do to improve performance I'd like to know.

meilisearch is running on thinkserver from 2016, 16gb ram and storage is on a SSD.

**Comment by ogizanagi** [comment]
Hi there 👋 
First of all, thanks to the Meili team and contributors for their wonderful work on this tool. 👏 
We're considering using it on many projects and started by a POC on a concrete case for one of our client, with great success.

However, I observed the same while indexing 2M+ documents, by batch of 2000 items:

![image](https://user-images.githubusercontent.com/2211145/89765238-3be69a80-daf6-11ea-8f06-3b339a2c1ca9.png)

_where duration is the one returned by the [updates API](https://docs.meilisearch.com/references/updates.html#get-an-update-status) for each insertion of 2000 documents_

So while sending API updates to the Meilisearch server for 2M items took about 2hours, the time for it being processed entirely is about a day in my case. It's even longer on a non-empty index when trying to update all documents, since the duration per update linearly grows with the number of already indexed documents.

What is even more strange, is the duration taken by an update seems uncorrelated with the number of documents inside:

### batch of 5000 documents:

```json
{
  "status": "processed",
  "updateId": 1054,
  "type": {
    "name": "DocumentsAddition",
    "number": 5000
  },
  "duration": 188.70902601,
  "enqueuedAt": "2020-08-05T07:22:36.835364801Z",
  "processedAt": "2020-08-05T07:25:45.805767057Z"
}
```

### batch of 10 documents:

```json
{
  "status": "processed",
  "updateId": 1055,
  "type": {
    "name": "DocumentsAddition",
    "number": 10
  },
  "duration": 165.732091704,
  "enqueuedAt": "2020-08-05T07:26:53.617585725Z",
  "processedAt": "2020-08-05T07:29:39.358857572Z"
}
```

**=> 10 or 5000 documents takes about the same time to be processed.**

This might be a real issue if we want to do "real-time" updates of our documents; sending such updates one by one may create a bottleneck quite fast, with more updates sent than capacity to process them.

(should I open another issue for this last observation?)

**Comment by Kerollmops** [comment]
Hello @gnur and @ogizanagi,

We now that one of the current bottlenec of the engine is on the indexing part already, we are currently working on a big refacto of the indexing system, it will be able to index in parallel and we keep much of the cores busy. The current benchmarks are really appeling as it can index more that 500m documents on 32cores on 2h.

If possible I would like to [know your settings](https://docs.meilisearch.com/references/settings.html#get-settings), this way we can track down what is taking so much time at indexing. 

Thank you verry much for those great measurements,
I which you a good day!

**Comment by ogizanagi** [comment]
Sure! 

<details>

<summary>Here are our setting for our offers index</summary>

```json
{
  "rankingRules": [
    "typo",
    "words",
    "proximity",
    "attribute",
    "wordsPosition",
    "exactness",
    "desc(onlineAt)",
    "desc(createdAt)",
    "desc(updatedAt)"
  ],
  "distinctAttribute": null,
  "searchableAttributes": [
    "objectID",
    "publicReference",
    "sellerReference",
    "publicID",
    "fullTitle",
    "dealerUrl",
    "sellerName",
    "sellerEmails",
    "manufacturers_names",
    "model",
    "company",
    "offerNotes",
    "sellerNotes",
    "description"
  ],
  "stopWords": [
    "a",
    "an",
    "in",
    "of",
    "the",
    "to"
  ],
  "synonyms": {},
  "acceptNewFields": false,
  "attributesForFaceting": [
    "company",
    "manufacturers_names",
    "category_lvl2",
    "country",
    "category_lvl1",
    "type",
    "categories",
    "category_lvl0",
    "zone"
  ]
}
```

</details>

**Comment by MCOfficer** [comment]
![Meilisearch Updates plotted](https://user-images.githubusercontent.com/22377202/96557563-b849e680-12ba-11eb-9eff-ed611e4b614d.png)


I'm seeing the same issue while indexing ~50M documents. (The missing updates towards the end failed due to hitting the LMDB size limit.)

Would it help to get a flamegraph of the indexing process?

**Comment by curquiza** [comment]
Hello @MCOfficer, @gnur and everyone following the issue!

The [first RC of MeiliSearch v0.21.0](https://github.com/meilisearch/MeiliSearch/releases/tag/v0.21.0rc0) is out.
We did our best to fix indexation and crash issues. We succeeded to improve it, but not to totally fix them.

You can test this new release by downloading the binaries available in this release.
Or you can use it with docker:

```bash
docker run -p 7700:7700 getmeili/meilisearch:v0.21.0rc0 ./meilisearch
```

We will still improve this after the release of the v0.21.0. We would rather release a non-completely optimized version rather than delay it and, at the same time, delay the release of new features. Be sure we are doing our best to always improve these indexation issues.

As a reminder:
* despite the improvements, we always recommend pushing your documents to MeiliSearch by batch and not one by one. The maximum payload size of MeiliSearch is 100Mb and [can be optionally increased](https://docs.meilisearch.com/reference/features/configuration.html#payload-limit-size). It means most of the dataset can be pushed with one call to the API.
* If you still have a memory crash, it means the RAM of your machine might not be adapted to your dataset size. We recommend increasing the RAM of your machine.

Thanks for your patience and your help with this! ❤️

**Comment by curquiza** [comment]
@guy4261  It means, by default, the size of your payload (not the number of documents) can by maximum 100Mb, so 100Mb or less :)

**Comment by curquiza** [comment]
Hello everyone reading this issue.

I close this issue due to the lack of feedback since it was happening with an old version of Meilisearch.

However, for anyone having issues in general with indexation performance, like this issue, please, refer to [this discussion](https://github.com/meilisearch/meilisearch/discussions/2131). Your feedback will be really appreciated 🙂
